File: Archive_LJ_isothermal_isobaric/znptmclj.c
/**********************************************************************
 *
 * File: znptmclj.c
 *
 * NpT-Monte Carlo of Lennard-Jonesium
 * Re-initialize checkpoint file
 *
 * 03-Apr-2010 (MN)
 * 17-Apr-2012
 *
 **********************************************************************/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "getval.h"

#define NL '\n'
#define BSIZE 80

/**********************************************************************/

int main() {

/**********************************************************************/

  FILE *fpi,*fpo;
  char fname[BSIZE];
  unsigned short seed[3];
  int n,nt,ntjob,ntprint,ntskip;
  double disp,dlnv,dr,p,pvm,t,v;
  double *x,*y,*z;

  /* Read old checkpoint file */

  strcpy(fname,"nptmclj_out.dat");
  printf("         infile=[nptmclj_out.dat] ");
  getval_s(fname);

  fpi=fopen(fname,"r");

  fread(&n,sizeof(int),1,fpi);
  fread(&nt,sizeof(int),1,fpi);
  fread(&ntjob,sizeof(int),1,fpi);
  fread(&ntprint,sizeof(int),1,fpi);
  fread(&ntskip,sizeof(int),1,fpi);
  fread(seed,sizeof(unsigned short),3,fpi);

  fread(&disp,sizeof(double),1,fpi);
  fread(&dlnv,sizeof(double),1,fpi);
  fread(&dr,sizeof(double),1,fpi);
  fread(&p,sizeof(double),1,fpi);
  fread(&pvm,sizeof(double),1,fpi);
  fread(&t,sizeof(double),1,fpi);
  fread(&v,sizeof(double),1,fpi);

  /* Allocate arrays */

  x=(double*)malloc(n*sizeof(double));
  y=(double*)malloc(n*sizeof(double));
  z=(double*)malloc(n*sizeof(double));

  /* Positions */

  fread(x,sizeof(double),n,fpi);
  fread(y,sizeof(double),n,fpi);
  fread(z,sizeof(double),n,fpi);

  fclose(fpi);

  /* User input */

  printf("              n=%16d\n",n);
  printf("              p=[%15.5lf] ",p);
  getval_d(&p);
  printf("              t=[%15.5lf] ",t);
  getval_d(&t);
  printf("           disp=[%15.5lf] ",disp);
  getval_d(&disp);
  printf("           dv/v=[%15.5lf] ",dlnv);
  getval_d(&dlnv);
  printf("    prob(vmove)=[%15.5lf] ",pvm);
  getval_d(&pvm);
  printf("             dr=[%15.5lf] ",dr);
  getval_d(&dr);
  printf("         ntskip=[%15d] ",ntskip);
  getval_i(&ntskip);
  printf(" ntprint/ntskip=[%15d] ",ntprint);
  getval_i(&ntprint);
  printf("          ntjob=[%15d] ",ntjob);
  getval_i(&ntjob);

  /* Write new startup file */

  strcpy(fname,"nptmclj_in.dat");
  printf("        outfile=[nptmclj_out.dat] ");
  getval_s(fname);

  nt=0;

  fpo=fopen(fname,"w");

  fwrite(&n,sizeof(int),1,fpo);
  fwrite(&nt,sizeof(int),1,fpo);
  fwrite(&ntjob,sizeof(int),1,fpo);
  fwrite(&ntprint,sizeof(int),1,fpo);
  fwrite(&ntskip,sizeof(int),1,fpo);
  fwrite(seed,sizeof(unsigned short),3,fpo);

  fwrite(&disp,sizeof(double),1,fpo);
  fwrite(&dlnv,sizeof(double),1,fpo);
  fwrite(&dr,sizeof(double),1,fpo);
  fwrite(&p,sizeof(double),1,fpo);
  fwrite(&pvm,sizeof(double),1,fpo);
  fwrite(&t,sizeof(double),1,fpo);
  fwrite(&v,sizeof(double),1,fpo);

  fwrite(x,sizeof(double),n,fpo);
  fwrite(y,sizeof(double),n,fpo);
  fwrite(z,sizeof(double),n,fpo);

  fclose(fpo);

  /* Deallocate arrays */

  free(x);
  free(y);
  free(z);

  return 0;

}

/**********************************************************************/

File: Archive_LJ_isothermal_isobaric/znptmclj.f90
!**********************************************************************!
!
! File: znptmclj.f90
!
! NpT-Monte Carlo of Lennard-Jonesium
! Re-initialize checkpoint file
!
! 01-May-1999 (MN)
! 17-Apr-2010
!
!**********************************************************************!

module getval_m

!**********************************************************************!

! Generic subroutine: read item from standard input/keep old value on
! hitting return

  implicit none

  private::getint,getreal,getstrg
  public::getval

  interface getval
    module procedure getint,getreal,getstrg
  end interface

  character(len=80),private::line

contains

!**********************************************************************!

  subroutine getint(intval)

!**********************************************************************!

    integer,intent(in out)::intval

    read(unit=*,fmt="(a)") line
    if(len_trim(adjustl(line))>0) then
      read(unit=line,fmt=*) intval
    end if

    return

  end subroutine getint

!**********************************************************************!

  subroutine getreal(realval)

!**********************************************************************!

    real,intent(in out)::realval

    read(unit=*,fmt="(a)") line
    if(len_trim(adjustl(line))>0) then
      read(unit=line,fmt=*) realval
    end if

    return

  end subroutine getreal

!**********************************************************************!

  subroutine getstrg(strgval)

!**********************************************************************!

    character(len=*),intent(in out)::strgval

    read(unit=*,fmt="(a)") line
    if(len_trim(adjustl(line))>0) then
      read(unit=line,fmt="(a)") strgval
    end if

    return

  end subroutine getstrg

end module getval_m

!**********************************************************************!

program znptmclj

!**********************************************************************!

  use getval_m

  implicit none

  integer,parameter::iin=1,iout=2

  character(len=80)::fname
  integer::n,nt,ntjob,ntprint,ntskip
  integer::nran
  integer,dimension(:),allocatable::iran
  real::disp,dlnv,dr,p,pvm,t,v
  real,dimension(:),allocatable::x,y,z

! RNG characteristics

  call random_seed(size=nran)

  allocate(iran(nran))

! Read old checkpoint file

  fname="nptmclj_out.dat"
  write(unit=*,fmt="(a)",advance="no") &
       "         infile=[nptmclj_out.dat] "
  call getval(fname)

  open(unit=iin,file=fname,status="old",action="read", &
       form="unformatted",position="rewind")

  read(unit=iin) n,nt,ntjob,ntprint,ntskip,iran
  read(unit=iin) disp,dlnv,dr,p,pvm,t,v

! Allocate arrays

  allocate(x(n),y(n),z(n))

! Positions

  read(unit=iin) x,y,z

  close(unit=iin)

! User input

  write(unit=*,fmt="(a,i16)") &
       "              n=",n
  write(unit=*,fmt="(a,f15.5,a)",advance="no") &
       "              p=[",p,"] "
  call getval(p)
  write(unit=*,fmt="(a,f15.5,a)",advance="no") &
       "              t=[",t,"] "
  call getval(t)
  write(unit=*,fmt="(a,f15.5,a)",advance="no") &
       "           disp=[",disp,"] "
  call getval(disp)
  write(unit=*,fmt="(a,f15.5,a)",advance="no") &
       "           dv/v=[",dlnv,"] "
  call getval(dlnv)
  write(unit=*,fmt="(a,f15.5,a)",advance="no") &
       "    prob(vmove)=[",pvm,"] "
  call getval(pvm)
  write(unit=*,fmt="(a,f15.5,a)",advance="no") &
       "             dr=[",dr,"] "
  call getval(dr)
  write(unit=*,fmt="(a,i15,a)",advance="no") &
       "         ntskip=[",ntskip,"] "
  call getval(ntskip)
  write(unit=*,fmt="(a,i15,a)",advance="no") &
       " ntprint/ntskip=[",ntprint,"] "
  call getval(ntprint)
  write(unit=*,fmt="(a,i15,a)",advance="no") &
       "   ntjob/ntskip=[",ntjob,"] "
  call getval(ntjob)

! Write new startup file

  fname="nptmclj_in.dat"
  write(unit=*,fmt="(a)",advance="no") &
       "        outfile=[ nptmclj_in.dat] "
  call getval(fname)

  nt=0

  open(unit=iout,file=fname,status="replace",action="write", &
       form="unformatted")

  write(unit=iout) n,nt,ntjob,ntprint,ntskip,iran
  write(unit=iout) disp,dlnv,dr,p,pvm,t,v
  write(unit=iout) x,y,z

  close(unit=iout)

! Deallocate arrays

  deallocate(iran,x,y,z)

end program znptmclj

!**********************************************************************!
